# Correlate distance with motifs #

Loop trough all the files and analyze the results.

First step: loading data
```{r}
# Load files of genes for each motifs of each species
extract_bi <- dir("../motifs_extraction", full.names=TRUE, pattern="bi.res$")
extract_ca <- dir("../motifs_extraction", full.names=TRUE, pattern="ca.res$")
extract_sex <- dir("../motifs_extraction", full.names=TRUE, pattern="sex.res$")
extract_tet <- dir("../motifs_extraction", full.names=TRUE, pattern="tet.res$")

# Load expression data
exp_bi = read.table("../../data/biaurelia_RNAseq_xp.tab", h=T)
exp_ca = read.table("../../data/caudatum_RNAseq_xp.tab", h=T)
exp_sex = read.table("../../data/sexaurelia_RNAseq_xp.tab", h=T)
exp_tet = read.table("../../data/tetraurelia_RNAseq_xp.tab", h=T)
```

Function to treat each files from each list
```{r}
# Function to use on every file, returns a table of matching genes between expression and extracted data
corr.dist.exp <- function(extract_file, expression_file) {
  extract_table = read.table(extract_file, h=F)
  colnames(extract_table) = c("gene_id", "distance", "motifs.number")
  extract_match = extract_table[match(expression_file$gene_id, extract_table$gene_id, nomatch=0),]
  extract.merge = merge(extract_match,expression_file)
}

# Function to returns p.value of fitted models
lm.pvalue <- function(table.merged) {
  model = lm(xp~distance, data=table.merged)  # Produce a linear model for data
  p.value = summary(model)$coefficients[[8]]
}
```

Test on *Paramecium biaurelia*, see the number of significant correlations.

```{r}
bi.pvalues = rep(NA,length(extract_bi))

for (i in 1:length(extract_bi)) {
  merged = corr.dist.exp(extract_bi[i], exp_bi)

  if (length(merged[,1])>2) {
    p.value = lm.pvalue(merged)
    bi.pvalues[i] = p.value
  }
}
length(bi.pvalues)
summary(bi.pvalues)
length(subset(bi.pvalues, bi.pvalues < 0.001))
```

For *P. caudatum*

```{r}
ca.pvalues = rep(NA,length(extract_ca))

for (i in 1:length(extract_ca)) {
  merged = corr.dist.exp(extract_ca[i], exp_ca)

  if (length(merged[,1])>2) {
    p.value = lm.pvalue(merged)
    ca.pvalues[i] = p.value
  }
}
length(ca.pvalues)
summary(ca.pvalues)
length(subset(ca.pvalues, ca.pvalues < 0.001))
```

For *P. sexaurelia*

```{r}
sex.pvalues = rep(NA,length(extract_sex))

for (i in 1:length(extract_sex)) {
  merged = corr.dist.exp(extract_sex[i], exp_sex)

  if (length(merged[,1])>2) {
    p.value = lm.pvalue(merged)
    sex.pvalues[i] = p.value
  }
}
length(sex.pvalues)
summary(sex.pvalues)
length(subset(sex.pvalues, sex.pvalues < 0.001))
```

For *P. tetraurelia*
```{r}
tet.pvalues = rep(NA,length(extract_tet))

for (i in 1:length(extract_tet)) {
  merged = corr.dist.exp(extract_tet[i], exp_tet)

  if (length(merged[,1])>2) {
    p.value = lm.pvalue(merged)
    tet.pvalues[i] = p.value
  }
}
length(tet.pvalues)
summary(tet.pvalues)
length(subset(tet.pvalues, tet.pvalues < 0.001))
```