# Correlate distance with motifs #

Loop trough all the files and analyze the results.

First step: loading data
```{r}
# Load files of genes for each motifs of each species
extract_bi <- dir("../motifs_extraction", full.names=TRUE, pattern="bi.res$")
extract_ca <- dir("../motifs_extraction", full.names=TRUE, pattern="ca.res$")
extract_sex <- dir("../motifs_extraction", full.names=TRUE, pattern="sex.res$")
extract_tet <- dir("../motifs_extraction", full.names=TRUE, pattern="tet.res$")

# Load expression data
exp_bi = read.table("../../data/biaurelia_RNAseq_xp.tab", h=T)
exp_ca = read.table("../../data/caudatum_RNAseq_xp.tab", h=T)
exp_sex = read.table("../../data/sexaurelia_RNAseq_xp.tab", h=T)
exp_tet = read.table("../../data/tetraurelia_RNAseq_xp.tab", h=T)

# Load motifs name <-> motif correspondance table
motifs = read.table("../15may14consensuscomp.txt", sep="\t")
head(motifs)
colnames(motifs) = c("family", "BF.motif", "BF.size", "MEME.motif", "MEME.size", "gene", "match.ratio", "consensus")
```

Function to treat each files from each list
```{r}
# Function to use on every file, returns a table of matching genes between expression and extracted data
corr.dist.exp <- function(extract_file, expression_file) {
  extract_table = read.table(extract_file, h=F)
  colnames(extract_table) = c("gene_id", "distance", "motifs.number")
  extract_match = extract_table[match(expression_file$gene_id, extract_table$gene_id, nomatch=0),]
  extract.merge = merge(extract_match,expression_file)
}

# Function to returns p.value of fitted models
lm.pvalue <- function(table.merged) {
  model = lm(xp~distance, data=table.merged)  # Produce a linear model for data
  p.value = summary(model)$coefficients[[8]]
  rsq = summary(model)$adj.r.squared
  list(p.value, rsq)
}
```

Test on *Paramecium biaurelia*, see the number of significant correlations.

```{r}
bi.pvalues = rep(NA,length(extract_bi))
bi.rsq = rep(NA, length(extract_bi))

for (i in 1:length(extract_bi)) {
  merged = corr.dist.exp(extract_bi[i], exp_bi)

  if (length(merged[,1])>2) {
    mod = lm.pvalue(merged)
    p.value = mod[[1]][1]
    rsq = mod[[2]][1]
    bi.pvalues[i] = p.value
    bi.rsq[i] = rsq
  }
}
length(bi.pvalues)
summary(bi.pvalues)
length(subset(bi.pvalues, bi.pvalues < 0.001))
```

For *P. caudatum*

```{r}
ca.pvalues = rep(NA,length(extract_ca))
ca.rsq = rep(NA, length(extract_ca))

for (i in 1:length(extract_ca)) {
  merged = corr.dist.exp(extract_ca[i], exp_ca)

  if (length(merged[,1])>2) {
    mod = lm.pvalue(merged)
    p.value = mod[[1]][1]
    rsq = mod[[2]][1]
    ca.pvalues[i] = p.value
    ca.rsq[i] = rsq
  }
}
length(ca.pvalues)
summary(ca.pvalues)
length(subset(ca.pvalues, ca.pvalues < 0.001))
```

For *P. sexaurelia*

```{r}
sex.pvalues = rep(NA,length(extract_sex))
sex.rsq = rep(NA, length(extract_sex))

for (i in 1:length(extract_sex)) {
  merged = corr.dist.exp(extract_sex[i], exp_sex)

  if (length(merged[,1])>2) {
    mod = lm.pvalue(merged)
    p.value = mod[[1]][1]
    rsq = mod[[2]][1]
    sex.pvalues[i] = p.value
    sex.rsq[i] = rsq
  }
}
length(sex.pvalues)
summary(sex.pvalues)
length(subset(sex.pvalues, sex.pvalues < 0.001))
```

For *P. tetraurelia*
```{r}
tet.pvalues = rep(NA,length(extract_tet))
tet.rsq = rep(NA, length(extract_tet))

for (i in 1:length(extract_tet)) {
  merged = corr.dist.exp(extract_tet[i], exp_tet)

  if (length(merged[,1])>2) {
    mod = lm.pvalue(merged)
    p.value = mod[[1]][1]
    rsq = mod[[2]][1]
    tet.pvalues[i] = p.value
    tet.rsq[i] = rsq
  }
}
length(tet.pvalues)
summary(tet.pvalues)
length(subset(tet.pvalues, tet.pvalues < 0.001))
```

Let see if there are some motifs in common.
```{r}

# To have indexes of elements
bi.match = match(subset(bi.pvalues, bi.pvalues < 0.001), bi.pvalues)
ca.match = match(subset(ca.pvalues, ca.pvalues < 0.001), ca.pvalues)
sex.match = match(subset(sex.pvalues, sex.pvalues < 0.001), sex.pvalues)
tet.match = match(subset(tet.pvalues, tet.pvalues < 0.001), tet.pvalues)

# Common elements
common = intersect(intersect(intersect(bi.match, ca.match), sex.match), tet.match)
length(common)
```

Let's extract the sequence of conserved motifs
```{r}
common.motifs = data.frame()
for (i in 1:length(common)) {
  index = common[i]
  filename = extract_bi[index]
  print(filename)
  filename = strsplit(filename, "/")
  filename = filename[[1]][3]
  fam = substr(filename, 1, 12)
  fam.motif = substr(filename, 13, 18)
  common.motifs = rbind(common.motifs, subset(motifs, family == fam & BF.motif == fam.motif))
}
```
Look at the scatterplots
```{r}
library(ggplot2)
for (i in 1:length(common)) {
  merged = corr.dist.exp(extract_bi[common[i]], exp_bi)
  ggplot(merged, aes(x = distance, y = xp)) + geom_point(alpha = 0.5, position = "jitter")
}
```