# Correlate distance with motifs #

Loop trough all the files and analyze the results.

First step: loading data
```{r}
setwd("~/Documents/ENS/M1/Stage/para/results/analysis/")

extraction <- function(patt) {
  extract <- dir("../motifs_extraction", full.names=TRUE, pattern=patt)
}

# Load files of genes for each motifs of each species
extract_bi <- extraction("bi.res$")
extract_ca <- extraction("ca.res$")
extract_sex <- extraction("sex.res$")
extract_tet <- extraction("tet.res$")

# Load expression data
exp_bi = read.table("../../data/biaurelia_RNAseq_xp.tab", h=T)
exp_ca = read.table("../../data/caudatum_RNAseq_xp.tab", h=T)
exp_sex = read.table("../../data/sexaurelia_RNAseq_xp.tab", h=T)
exp_tet = read.table("../../data/tetraurelia_RNAseq_xp.tab", h=T)

# Load motifs name <-> motif correspondance table
motifs = read.table("../june1714consensuscomp.txt", sep="\t", h=T)
head(motifs)
colnames(motifs) = c("family", "BF.motif", "BF.size", "MEME.motif", "MEME.size", "gene", "match.ratio", "consensus")

# pvalue threshold to select in the whole file
pval.thre = 0.01
```

Function to treat each files from each list
```{r}
# Function to use on every file, returns a table of matching genes between expression and extracted data
corr.dist.exp <- function(extract_table, expression_file) {
  # Select only gene ids that are in expression file
  extract_match = subset(extract_table, gene_id %in% intersect(expression_file$gene_id, extract_table$gene_id))
  extract.merge = merge(extract_match,expression_file)
}

# Function to returns p.value of fitted models
lm.pvalue <- function(table.merged) {
  model = lm(xp~distance, data=table.merged)  # Produce a linear model for data
  p.value = summary(model)$coefficients[[8]]
  rsq = summary(model)$adj.r.squared
  list(p.value, rsq)
}
```

Test on *Paramecium biaurelia*, see the number of signifisexnt correlations.

```{r}
extract_bi_table = read.table(extract_bi)
colnames(extract_bi_table) = c("motif", "gene_id", "distance", "motifs.number")
bi_motifs = unique(extract_bi_table$motif)
bi.pvalues = rep(NA,length(bi_motifs))
bi.rsq = rep(NA, length(bi_motifs))
for (i in 1:length(bi_motifs)) {
  extract_bi_sub = subset(extract_bi_table, motif == bi_motifs[i])
  merged = corr.dist.exp(extract_bi_sub, exp_bi)
  if (length(merged[,1])>2) {
  mod = lm.pvalue(merged)
  p.value = mod[[1]][1]
  rsq = mod[[2]][1]
  bi.pvalues[i] = p.value
  bi.rsq[i] = rsq
  }
}

length(bi.pvalues)
summary(bi.pvalues)
length(subset(bi.pvalues, bi.pvalues < pval.thre))
```

For *P. caudatum*

```{r}
extract_ca_table = read.table(extract_ca)
colnames(extract_ca_table) = c("motif", "gene_id", "distance", "motifs.number")
ca_motifs = unique(extract_ca_table$motif)
ca.pvalues = rep(NA,length(ca_motifs))
ca.rsq = rep(NA, length(ca_motifs))
for (i in 1:length(ca_motifs)) {
  extract_ca_sub = subset(extract_ca_table, motif == ca_motifs[i])
  merged = corr.dist.exp(extract_ca_sub, exp_ca)
  if (length(merged[,1])>2) {
  mod = lm.pvalue(merged)
  p.value = mod[[1]][1]
  rsq = mod[[2]][1]
  ca.pvalues[i] = p.value
  ca.rsq[i] = rsq
  }
}

length(ca.pvalues)
summary(ca.pvalues)
length(subset(ca.pvalues, ca.pvalues < pval.thre))
```

For *P. sexaurelia*

```{r}
extract_sex_table = read.table(extract_sex)
colnames(extract_sex_table) = c("motif", "gene_id", "distance", "motifs.number")
sex_motifs = unique(extract_sex_table$motif)
sex.pvalues = rep(NA,length(sex_motifs))
sex.rsq = rep(NA, length(sex_motifs))
for (i in 1:length(sex_motifs)) {
  extract_sex_sub = subset(extract_sex_table, motif == sex_motifs[i])
  merged = corr.dist.exp(extract_sex_sub, exp_sex)
  if (length(merged[,1])>2) {
  mod = lm.pvalue(merged)
  p.value = mod[[1]][1]
  rsq = mod[[2]][1]
  sex.pvalues[i] = p.value
  sex.rsq[i] = rsq
  }
}

length(sex.pvalues)
summary(sex.pvalues)
length(subset(sex.pvalues, sex.pvalues < pval.thre))
```

For *P. tetraurelia*
```{r}
extract_tet_table = read.table(extract_tet)
colnames(extract_tet_table) = c("motif", "gene_id", "distance", "motifs.number")
tet_motifs = unique(extract_tet_table$motif)
tet.pvalues = rep(NA,length(tet_motifs))
tet.rsq = rep(NA, length(tet_motifs))
for (i in 1:length(tet_motifs)) {
  extract_tet_sub = subset(extract_tet_table, motif == tet_motifs[i])
  merged = corr.dist.exp(extract_tet_sub, exp_tet)
  if (length(merged[,1])>2) {
  mod = lm.pvalue(merged)
  p.value = mod[[1]][1]
  rsq = mod[[2]][1]
  tet.pvalues[i] = p.value
  tet.rsq[i] = rsq
  }
}

length(tet.pvalues)
summary(tet.pvalues)
length(subset(tet.pvalues, tet.pvalues < pval.thre))
```

Let see if there are some motifs in common.
```{r}
# To have indexes of elements
bi.match = match(subset(bi.pvalues, bi.pvalues < pval.thre), bi.pvalues)
ca.match = match(subset(ca.pvalues, ca.pvalues < pval.thre), ca.pvalues)
sex.match = match(subset(sex.pvalues, sex.pvalues < pval.thre), sex.pvalues)
tet.match = match(subset(tet.pvalues, tet.pvalues < pval.thre), tet.pvalues)

# Common elements
common = intersect(intersect(intersect(bi_motifs[bi.match], ca_motifs[ca.match]), sex_motifs[sex.match]), tet_motifs[tet.match])
length(common)
```

Let's extract the sequence of conserved motifs
```{r}
common.motifs = data.frame()

for (i in 1:length(common)) {
  index = common[i]
  filename = extract_bi[index]
  print(filename)
  filename = strsplit(filename, "/")
  filename = filename[[1]][3]
  fam = substr(filename, 1, 12)
  fam.motif = substr(filename, 13, 18)
  common.motifs = rbind(common.motifs, subset(motifs, family == fam & BF.motif == fam.motif))
}
common.motifs
```

Look at the scatterplots
```{r}
library(ggplot2)
for (i in 1:length(common)) {
  merged = corr.dist.exp(extract_bi[common[i]], exp_bi)
  index = common[i]
  filename = extract_bi[index]
  filename = strsplit(filename, "/")
  filename = filename[[1]][3]
  fam = substr(filename, 1, 12)
  fam.motif = substr(filename, 13, 18)
  mot = subset(common.motifs, family == fam & BF.motif == fam.motif)$consensus
  print(ggplot(merged, aes(x = distance, y = xp)) + geom_point(alpha = 0.5, position = "jitter") + labs(x="Distance from Start Codon", y="Gene Expression", title = paste("Distance - Expression for", mot)))
}
```

If we try to classify motifs as associated with high or low expression among species?
The idea is two try to see if genes with given motif are higher or lower expressed than genes without.

```{r}
# Function to classify motifs as increaser or decreaser of gene expression
expression.modifier <- function(extract_table, exp_file) {
  # Expression of genes that contain the motif
  extract_match = exp_file[match(extract_table$gene_id, exp_file$gene_id, nomatch=0),] 
  # Expression of genes without the motif
  extract_no = exp_file[!match(exp_file$gene_id, extract_table$gene_id, nomatch=0),]
  # Compare if the means, null hypothesis: means are equal (better than t-test besexuse non-normal distribution)
  w.test = wilcox.test(extract_match$xp, extract_no$xp)
  if (w.test$p.value < pval.thre) {
    
    # If genes with motif are more expressed, the motif is increasing expression
    if (mean(extract_match$xp > mean(extract_no$xp))) {
      classifier = "high"
    } else {
      classifier = "low"
    }
  }
  else {
    classifier = NA
  }
  list(mod=classifier, match=length(extract_match[,1]), nomatch=length(extract_no[,1]), avg.dist=mean(extract_table$distance))
}
```


```{r}
bi.classify = data.frame()

for (i in 1:length(bi_motifs)) {
  extract_bi_sub = subset(extract_bi_table, motif == bi_motifs[i])
  
  modifier = expression.modifier(extract_bi_sub, exp_bi)
  motif_data = subset(motifs, as.character(consensus) == as.character(bi_motifs[i]))
  bi.classify = rbind(bi.classify ,cbind(motif_data, as.factor(modifier$mod), modifier$match, modifier$nomatch,modifier$avg.dist))
  
}
colnames(bi.classify)[9:11]= c("modifier", "bi.match", "bi.nomatch")
```

```{r}
ca.classify = data.frame()

for (i in 1:length(ca_motifs)) {
  extract_ca_sub = subset(extract_ca_table, motif == ca_motifs[i])
  
  modifier = expression.modifier(extract_ca_sub, exp_ca)
  motif_data = subset(motifs, as.character(consensus) == as.character(ca_motifs[i]))
  ca.classify = rbind(ca.classify ,cbind(motif_data, as.factor(modifier$mod), modifier$match, modifier$nomatch,modifier$avg.dist))
  
}
colnames(ca.classify)[9:11]= c("modifier", "ca.match", "ca.nomatch")
```

```{r}
sex.classify = data.frame()

for (i in 1:length(sex_motifs)) {
  extract_sex_sub = subset(extract_sex_table, motif == sex_motifs[i])
  
  modifier = expression.modifier(extract_sex_sub, exp_sex)
  motif_data = subset(motifs, as.character(consensus) == as.character(sex_motifs[i]))
  sex.classify = rbind(sex.classify ,cbind(motif_data, as.factor(modifier$mod), modifier$match, modifier$nomatch,modifier$avg.dist))
  
}
colnames(sex.classify)[9:11]= c("modifier", "sex.match", "sex.nomatch")
```

```{r}
tet.classify = data.frame()

for (i in 1:length(tet_motifs)) {
  extract_tet_sub = subset(extract_tet_table, motif == tet_motifs[i])
  
  modifier = expression.modifier(extract_tet_sub, exp_tet)
  motif_data = subset(motifs, as.character(consensus) == as.character(tet_motifs[i]))
  tet.classify = rbind(tet.classify ,cbind(motif_data, as.factor(modifier$mod), modifier$match, modifier$nomatch,modifier$avg.dist))
  
}
colnames(tet.classify)[9:11]= c("modifier", "tet.match", "tet.nomatch")
```

Let's look at coregulation patterns in all the species

```{r}
library(scales)
common.regulation = cbind(subset(bi.classify, consensus %in% common)[8:11], subset(ca.classify, consensus %in% common)[9:11], subset(sex.classify, consensus %in% common)[9:11], subset(tet.classify, consensus %in% common)[9:11])
colnames(common.regulation)[c(2,5,8,11)] = c("bi.mod", "ca.mod", "sex.mod", "tet.mod")
common.regulation = common.regulation[!duplicated(common.regulation$consensus),]
common.regulation
library(reshape2)
common.melt = melt(common.regulation)[,c(1, 6, 7)]
common.tot = melt(common.regulation)
com = cbind(common.melt, colsplit(common.melt$variable, names=c("species", "type"), pattern="\\."))
common.cast = dcast(com, consensus+species~type)

common.tot = cbind(common.tot, colsplit(common.tot$variable, names=c("species", "type"), pattern="\\."))
tot.cast = dcast(common.tot, consensus+bi.mod+ca.mod+sex.mod+tet.mod~type+species)

c = cbind(common.cast, common.cast$match/(common.cast$match+common.cast$nomatch))
colnames(c)[5] = "ratio"

reduced = dcast(melt(c), consensus~species+variable)[,c(1,2,4,5,7,8,10,11,13)]
reduced = cbind(reduced, tot.cast$bi.mod, tot.cast$ca.mod, tot.cast$sex.mod, tot.cast$tet.mod)
colnames(reduced)[10:13] = c("bi_mod", "ca_mod", "sex_mod", "tet_mod")
reduced[,c(3,5,7,9)] = round(reduced[,c(3,5,7,9)]*100,2)

# Each motif separated
ggplot(c, aes(x=species, y=ratio)) +
  geom_point() +
  labs(x="Species", y="Proportion of genes with motifs") +
  facet_wrap(~consensus, ncol=5) +
  scale_y_continuous(labels=percent) +
  theme(text=element_text(size=rel(4.5)))

# All motifs by species
ggplot(c, aes(x=species, y=ratio, fill=species))+
  geom_boxplot() +
  geom_jitter() +
  labs(x="Species", y="Proportion of genes with motifs") +
  scale_fill_discrete(labels=c("P. biaurelia", "P. caudatum", "P. sexaurelia", "P. tetraurelia")) +
  scale_y_continuous(labels=percent) +
  theme(text=element_text(size=rel(4.5)), axis.text.x=element_text(size=rel(5)), legend.position="none")

# Group Motifs
ggplot(c, aes(x=consensus, y=ratio)) +
  geom_boxplot() +
  geom_jitter() +
  labs(x="Motifs", y="Proportion of genes with motifs") +
  scale_y_continuous(labels=percent) +
  theme(legend.position="none", text=element_text(size=rel(4.5)), axis.text.x=element_text(size=rel(4.5)))
```

GLM
```{r}
library(car)
# With all sequences
for (i in 1:length(unique(common.cast$consensus))) {
  start = 1+(i-1)*4
  end = 4+(i-1)*4
  print(i)
  glm = glm(cbind(match,nomatch)~species, family=binomial, common.cast[start:end,])
  print(Anova(glm))
}

# Without caudatum
for (i in 1:length(unique(common.cast$consensus))) {
  start = 1+(i-1)*4
  end = 4+(i-1)*4
  print(i)
  sub = subset(common.cast[start:end,], species != "ca")
  glm = glm(cbind(match,nomatch)~species, family=binomial, common.cast[start:end,])
  print(Anova(glm))
}
```

```{r}
val.common = data.frame(bi.pvalues[common], bi.rsq[common], ca.pvalues[common], ca.rsq[common], sex.pvalues[common], sex.rsq[common], tet.pvalues[common], tet.rsq[common])
colnames(val.common) = c("bi.pval", "bi.rsq", "ca.pval", "ca.rsq", "sex.pval", "sex.rsq", "tet.pval", "tet.rsq")
```